File: ARTED/stencil/C/AVX/current.c
/*
 *  Copyright 2016 ARTED developers
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

/* Hand-Code Vector processing for Intel Ivy-Bridge (AVX) */

#include <complex.h>
#include "../interop.h"

extern int NLx, NLy, NLz;

#ifndef ARTED_DOMAIN_POWER_OF_TWO
extern int *modx, *mody, *modz;
#endif

#ifdef ARTED_STENCIL_LOOP_BLOCKING
extern int BX, BY;
#endif

void current_stencil_( double         const           C[restrict 12]
                     , double complex const           E[restrict NLx][NLy][NLz]
                     , double              * restrict F
                     , double              * restrict G
                     , double              * restrict H
) {
  double complex const* e;

  int ix, iy, iz, n;

#ifdef ARTED_STENCIL_LOOP_BLOCKING
  int bx, by;
#endif

  const __m256i CONJ = _mm256_set_epi64x(1LL << 63, 0, 1LL << 63, 0);

  __m256d tx = _mm256_setzero_pd();
  __m256d ty = _mm256_setzero_pd();
  __m256d tz = _mm256_setzero_pd();

#pragma novector
#ifdef ARTED_STENCIL_LOOP_BLOCKING
  for(bx = 0 ; bx < NLx ; bx += BX)
  for(by = 0 ; by < NLy ; by += BY)
  for(ix = bx ; ix < MIN(bx+BX,NLx) ; ++ix) {
  for(iy = by ; iy < MIN(by+BY,NLy) ; ++iy)
#else
  for(ix = 0 ; ix < NLx ; ++ix) {
  for(iy = 0 ; iy < NLy ; ++iy)
#endif
  {
    e = &E[ix][iy][0];

    for(iz = 0 ; iz < NLz ; iz += 2)
    {
#define STENCIL_CALC(PP,CC) \
      v2 = _mm256_mul_pd(_mm256_broadcast_sd(CC), PP); \
      v1 = _mm256_add_pd(v1, v2);

#define STENCIL(IDP,CC) \
      p = _mm256_load_pd((double *) (e + iz - IDP));     \
      STENCIL_CALC(p,CC) \

      __m256d m, p;
      __m256d v1, v2, v3, v4;

      // conj(e[iz])
      __m256d ez = _mm256_load_pd((double *)(e + iz));
      __m256d w = (__m256d) _mm256_xor_si256((__m256i) ez, CONJ);

      /* x-dimension (NLy*NLz stride)  */
      {
        v1 = _mm256_setzero_pd();
        STENCIL(IDX(1), C+0);
        STENCIL(IDX(2), C+1);
        STENCIL(IDX(3), C+2);
        STENCIL(IDX(4), C+3);
        v3 = _mm256_shuffle_pd(v1, v1, 0x05);
        v4 = _mm256_mul_pd(w, v3);
        tx = _mm256_add_pd(v4, tx);
      }

      /* y-dimension (NLz stride) */
      {
        v1 = _mm256_setzero_pd();
        STENCIL(IDY(1), C+4);
        STENCIL(IDY(2), C+5);
        STENCIL(IDY(3), C+6);
        STENCIL(IDY(4), C+7);
        v3 = _mm256_shuffle_pd(v1, v1, 0x05);
        v4 = _mm256_mul_pd(w, v3);
        ty = _mm256_add_pd(v4, ty);
      }

      /* z-dimension (unit stride) */
      {
        __m256d z2,z3,z5,z7;
#ifdef ARTED_DOMAIN_POWER_OF_TWO
        z2 = _mm256_load_pd((double *)(e + ((iz + 2 + NLz) & (NLz - 1))));
        z3 = _mm256_load_pd((double *)(e + ((iz + 4 + NLz) & (NLz - 1))));
#else
        z2 = _mm256_load_pd((double *)(e + modz[iz + 2 + NLz]));
        z3 = _mm256_load_pd((double *)(e + modz[iz + 4 + NLz]));
#endif
        z5 = _mm256_permute2f128_pd(ez, z2, 0x21);
        z7 = _mm256_permute2f128_pd(z2, z3, 0x21);

        v1 = _mm256_setzero_pd();
        STENCIL_CALC(z2, C+ 9);
        STENCIL_CALC(z3, C+11);
        STENCIL_CALC(z5, C+ 8);
        STENCIL_CALC(z7, C+10);
        v3 = _mm256_shuffle_pd(v1, v1, 0x05);
        v4 = _mm256_mul_pd(w, v3);
        tz = _mm256_add_pd(v4, tz);
      }
    }  /* NLz */
  } /* NLy */
  } /* NLx */

  const __m256d two = _mm256_set1_pd(2);

  tx = _mm256_mul_pd(tx, two);
  ty = _mm256_mul_pd(ty, two);
  tz = _mm256_mul_pd(tz, two);

  __declspec(align(32)) double r[4];

  _mm256_store_pd(r, tx);
  *F = r[0] + r[1] + r[2] + r[3];

  _mm256_store_pd(r, ty);
  *G = r[0] + r[1] + r[2] + r[3];

  _mm256_store_pd(r, tz);
  *H = r[0] + r[1] + r[2] + r[3];
}

File: ARTED/stencil/F90/current.f90
!
!  Copyright 2016 ARTED developers
!
!  Licensed under the Apache License, Version 2.0 (the "License");
!  you may not use this file except in compliance with the License.
!  You may obtain a copy of the License at
!
!      http://www.apache.org/licenses/LICENSE-2.0
!
!  Unless required by applicable law or agreed to in writing, software
!  distributed under the License is distributed on an "AS IS" BASIS,
!  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
!  See the License for the specific language governing permissions and
!  limitations under the License.
!
subroutine current_stencil(C,E,F,G,H)
  use global_variables, only: NLx,NLy,NLz,zI
#ifdef ARTED_STENCIL_ENABLE_LOOP_BLOCKING
  use opt_variables, only: LBX => STENCIL_BLOCKING_X, LBY => STENCIL_BLOCKING_Y
#endif
#ifndef ARTED_DOMAIN_POWER_OF_TWO
  use opt_variables, only: modx, mody, modz
#endif
  implicit none
  real(8),    intent(in)  :: C(12)
  complex(8), intent(in)  :: E(0:NLz-1,0:NLy-1,0:NLx-1)
  real(8),    intent(out) :: F,G,H

#ifdef ARTED_STENCIL_ENABLE_LOOP_BLOCKING
  integer    :: bx,by
#endif
  integer    :: ix,iy,iz
  complex(8) :: v,w

#ifdef __INTEL_COMPILER
#if defined(__KNC__) || defined(__AVX512F__)
#   define MEM_ALIGN   64
#   define VECTOR_SIZE 4
# else
#   define MEM_ALIGN   32
#   define VECTOR_SIZE 2
# endif

!dir$ assume_aligned E:MEM_ALIGN
#endif

#ifdef ARTED_DOMAIN_POWER_OF_TWO
# ifdef __INTEL_COMPILER
!dir$ assume (mod(NLx, VECTOR_SIZE) == 0)
!dir$ assume (mod(NLy, VECTOR_SIZE) == 0)
!dir$ assume (mod(NLz, VECTOR_SIZE) == 0)
# endif
# define IDX(dt) iz,iy,and(ix+(dt)+NLx,NLx-1)
# define IDY(dt) iz,and(iy+(dt)+NLy,NLy-1),ix
# define IDZ(dt) and(iz+(dt)+NLz,NLz-1),iy,ix
#else
# define IDX(dt) iz,iy,modx(ix+(dt)+NLx)
# define IDY(dt) iz,mody(iy+(dt)+NLy),ix
# define IDZ(dt) modz(iz+(dt)+NLz),iy,ix
#endif

  H = 0
  G = 0
  F = 0

#ifdef ARTED_STENCIL_ENABLE_LOOP_BLOCKING
  do bx=0,NLx-1,LBX
  do by=0,NLy-1,LBY
  do ix=bx,min(bx+LBX-1,NLx-1)
  do iy=by,min(by+LBY-1,NLy-1)
#else
  do ix=0,NLx-1
  do iy=0,NLy-1
#endif
#ifdef __INTEL_COMPILER
!dir$ simd
#endif
#ifdef __FUJITSU
!OCL simd
!OCL noalias
#endif
  do iz=0,NLz-1
    w = conjg(E(iz,iy,ix))

    v=(C( 9)*E(IDZ(1)) &
    & +C(10)*E(IDZ(2)) &
    & +C(11)*E(IDZ(3)) &
    & +C(12)*E(IDZ(4)))

    H = H + imag(w * v)

    v=(C( 5)*E(IDY(1)) &
    & +C( 6)*E(IDY(2)) &
    & +C( 7)*E(IDY(3)) &
    & +C( 8)*E(IDY(4)))

    G = G + imag(w * v)

    v=(C( 1)*E(IDX(1)) &
    & +C( 2)*E(IDX(2)) &
    & +C( 3)*E(IDX(3)) &
    & +C( 4)*E(IDX(4)))

    F = F + imag(w * v)
  end do
  end do
  end do
#ifdef ARTED_STENCIL_ENABLE_LOOP_BLOCKING
  end do
  end do
#endif
  H = H * 2.0d0
  G = G * 2.0d0
  F = F * 2.0d0
end subroutine
